### PR TITLE
json-c 0.14

### PR DESCRIPTION
**Pull request by chenrui333**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by chenrui333** [comment]
```
==> ./configure --disable-dependency-tracking --disable-silent-rules --prefix=/usr/local/Cellar/json-c/0.14
Failed to execute: ./configure
```

**Comment by EricFromCanada** [comment]
- libu2f-server fails due to [removal of TRUE / FALSE](https://github.com/json-c/json-c/blob/master/ChangeLog) defines; [fix is coming](https://github.com/Yubico/libu2f-server/pull/42)
- same for ykpers; [fix proposed](https://github.com/Yubico/yubikey-personalization/pull/159)
- dnsperf throws "BIND 9 libraries must be installed" on configure; appears unrelated
- poppler's site certificate has expired, preventing source downloads
- also failing: gdal, clamav

**Comment by micahsnyder** [comment]
>     * clamav fails on link due to missing symbols _iconv, _iconv_close, _iconv_open

I have an update for clamav (https://github.com/Homebrew/homebrew-core/pull/54096) that among other things also resolves the iconv symbols issue.

**Comment by chenrui333** [comment]
> dnsperf throws "BIND 9 libraries must be installed" on configure; appears unrelated

That is dnsperf compatibility issue with bind 9.16

**Comment by SMillerDev** [comment]
rebased everything to see if it works now

**Comment by SMillerDev** [comment]
```
brew install --verbose --build-bottle clamav
brew install --verbose --build-bottle dnsperf
brew linkage --test postgis
brew install --verbose --build-bottle mpw
brew install --verbose --build-bottle newsboat
brew install --verbose --build-bottle tracebox
brew audit --skip-style tracebox --online --git
brew install --verbose --build-bottle zmap
```

**Comment by SMillerDev** [comment]
- brew install --verbose --build-bottle dnsperf
```
   "_krb5_get_default_realm", referenced from:
      _check_config in libdns.a(gssapictx.o)
  "_krb5_init_context", referenced from:
      _check_config in libdns.a(gssapictx.o)
ld: symbol(s) not found for architecture x86_64
```

- brew install --verbose --build-bottle mpw
not compatible with json-c 0.14 yet it seems
```
INFO:     Enabled mpw_json (libjson-c).
core/mpw-marshal-util.c:97:53: error: use of undeclared identifier 'TRUE'
    return json_object_get_boolean( json_value ) == TRUE;
```

- brew install --verbose --build-bottle tracebox
I don't know why it can't find this
```
 checking for libpcap... no
configure: error: libpcap not found
```

- brew install --verbose --build-bottle zmap
```
 Undefined symbols for architecture x86_64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
Undefined symbols for architecture x86_64:
```

**Comment by Minoru** [comment]
JFYI Newsboat 2.20 fails to install on BSD because `install -D` have different meaning in GNU and BSD. This is fixed by https://github.com/newsboat/newsboat/pull/1011, and I will push out a 2.20.1 in three days (just waiting to see if any more critical bugs pop up).

File: Formula/clamav.rb [hunk 1]
```diff
@@ -44,7 +45,7 @@ def install
       --with-llvm=no
       --with-libiconv-prefix=#{Formula["libiconv"].opt_prefix}
       --with-iconv=#{Formula["libiconv"].opt_prefix}
-      --with-libjson-static=#{Formula["json-c"].opt_prefix}/lib/libjson-c.a
+      --with-libjson=#{Formula["json-c"].opt_prefix}
```
Comment (micahsnyder) [inline Formula/clamav.rb:48]: Can I assume this change was done because of a build error in clamav when using the json-c built by this recipe?  Perhaps we can revert this change in the next json-c release (0.15).  I forgot that the cmake tooling for 0.14 doesn't build with `-fPIC`.  It should be fixed in the next release (see https://github.com/json-c/json-c/issues/508)

Comment (SMillerDev) [reply Formula/clamav.rb:48]: No, it's because it doesn't provide static libraries by default anymore and homebrew prefers dynamic libraries.

**Comment by miccal** [comment]
> * brew install --verbose --build-bottle mpw
>   not compatible with json-c 0.14 yet it seems
> 
> ```
> INFO:     Enabled mpw_json (libjson-c).
> core/mpw-marshal-util.c:97:53: error: use of undeclared identifier 'TRUE'
>     return json_object_get_boolean( json_value ) == TRUE;
> ```

Changing `ENV["mpw_json"] = "1"` to `ENV["mpw_json"] = "0"` in `mpw` will disable JSON configuration support, which may help.

**Comment by SMillerDev** [comment]
That leaves just dnsperf broken on some weird krb5 issue and zmap on an even weirder issue.

File: Formula/clamav.rb (hunk 1 above)
Comment (micahsnyder) [reply Formula/clamav.rb:48]: Perhaps that stance on dynamic libraries could change in the future? Here's a pretty compelling argument why static libraries should be preferred over dynamic https://twitter.com/Cor3ntin/status/1276853471548002304

In this case, while the static json-c 0.14 library won't build in libclamav without the fix I mentioned to add `-fPIC` to CFLAGS, there are stability concerns when including the dynamic json-c library.  json-c shares symbol names with jansson, meaning applications that use json-c via a library like libclamav and also use jansson for their own json needs will see symbol collisions (i.e. random crashes) because the application has to link with both json-c and jansson dynamic libraries.  For this reason, I would recommend including libjson-c.a in the homebrew package when json-c 0.15 comes out.  For 0.14 it doesn't make as much difference [to clamav] because static linking to a .so will fail w/out `-fPIC`.

Comment (SMillerDev) [reply Formula/clamav.rb:48]: Json-c removed it from their default install, while we could add it back we follow upstream's defaults unless those are known to be insufficient. You should instead be asking json-c to revert their change.

Comment (micahsnyder) [reply Formula/clamav.rb:48]: It looks like someone has already addressed this a couple months ago for the next release: https://github.com/json-c/json-c/commit/e97fc20bfd99d341174dc61c30ce7f9540786a6f

**Comment by chenrui333** [comment]
I will do the rebase later (after we closed the openjdk PR)

**Comment by chenrui333** [comment]
side note, we have some major work around rust and sqlite to merge too.

**Comment by chenrui333** [comment]
10.15
```
Error: 5 failed steps!
brew install --verbose --build-bottle dnsperf
brew install --verbose --build-bottle mpw
brew install --verbose --build-bottle tracebox
brew audit --skip-style tracebox --online --git
brew install --verbose --build-bottle zmap
```

10.14
```
Error: 5 failed steps!
brew install --verbose --build-bottle dnsperf
brew install --verbose --build-bottle mpw
brew install --verbose --build-bottle tracebox
brew audit --skip-style tracebox --online --git
brew install --verbose --build-bottle zmap
```

10.13
```
Error: 4 failed steps!
brew install --verbose --build-bottle dnsperf
brew install --verbose --build-bottle mpw
brew audit --skip-style tracebox --online --git
brew install --verbose --build-bottle zmap
```

**Comment by chenrui333** [comment]
- `dnsperf`
- `mpw`
- `zmap`

**Comment by miccal** [comment]
For `mpw` see my comment [above](https://github.com/Homebrew/homebrew-core/pull/53407#issuecomment-650608888).

**Comment by chenrui333** [comment]
mpw got fixed with @miccal 's suggestion
dnsperf got fixed with @SeekingMeaning's LDFLAGS fix

Currently stuck with tracebox and zmap.

**Comment by SMillerDev** [comment]
Does zmap build successfully without json-c change?

**Comment by chenrui333** [comment]
> Does zmap build successfully without json-c change?

Yes, zmap build successfully with json-c 0.13.1 and I did file a bug report on the 0.14 compatibility issue in [here](https://github.com/zmap/zmap/issues/614).

**Comment by SMillerDev** [comment]
There is a commit in the repo that's supposed to fix it.

**Comment by chenrui333** [comment]
that sounds about right, the HEAD installation works for me.

**Comment by chenrui333** [comment]
zmap is fixed, now it is only tracebox has issue.

**Comment by chenrui333** [comment]
only tracebox is an issue at this moment.

**Comment by SMillerDev** [comment]
Lets go without tracebox.

**Comment by chenrui333** [comment]
Merging the PR now. 🚀

**Comment by chenrui333** [comment]
due to the license commits, i have to rebase my PR again. :(
